stacks-signer/src/cli.rs

```diff
@@ -123,6 +132,89 @@ pub struct RunSignerArgs {
     pub config: PathBuf,
 }
 
+#[derive(Parser, Debug, Clone, Copy)]
+/// Arguments for the Vote command
+pub struct GenerateVoteArgs {
+    /// The Stacks private key to use in hexademical format
+    #[arg(short, long, value_parser = parse_private_key)]
+    pub private_key: StacksPrivateKey,
+    /// The vote info being cast
+    #[clap(flatten)]
+    pub vote_info: VoteInfo,
+}
+
+#[derive(Parser, Debug, Clone, Copy)]
+/// Arguments for the VerifyVote command
+pub struct VerifyVoteArgs {
+    /// The Stacks public key to verify against
+    #[arg(short, long, value_parser = parse_public_key)]
+    pub public_key: StacksPublicKey,
+    /// The message signature in hexadecimal format
+    #[arg(short, long, value_parser = parse_message_signature)]
+    pub signature: MessageSignature,
+    /// The vote info being verified
+    #[clap(flatten)]
+    pub vote_info: VoteInfo,
+}
+
+#[derive(Parser, Debug, Clone, Copy)]
+/// Information about a SIP vote
+pub struct VoteInfo {
+    /// The SIP number to vote on
+    #[arg(long)]
+    pub sip: u32,
+    /// The vote to cast
+    #[arg(long, value_parser = parse_vote)]
+    pub vote: Vote,
+}
+
+impl VoteInfo {
+    /// Get the digest to sign that authenticates this vote data
+    fn digest(&self) -> Sha512Trunc256Sum {
+        let mut hasher = Sha512_256::new();
+        hasher.update(self.sip.to_be_bytes());
+        hasher.update((self.vote as u8).to_be_bytes());
+        Sha512Trunc256Sum::from_hasher(hasher)
+    }
+
+    /// Sign the vote data and return the signature
+    pub fn sign(&self, private_key: &StacksPrivateKey) -> Result<MessageSignature, &'static str> {
+        let digest = self.digest();
+        private_key.sign(digest.as_bytes())
+    }
```

Comment:
This message format is fine, but I think it'd be better if it were a SIP-018 signing scheme. A [SIP-018](https://github.com/stacksgov/sips/blob/main/sips/sip-018/sip-018-signed-structured-data.md#formal-specification) signing scheme is basically something like:

```rust
use clarity::vm::Value as ClarityValue;
use clarity::vm::types::TupleData;
use stackslib::util_lib::sign_structured_data::make_structured_data_domain;
use stackslib::util_lib::sign_structured_data::structured_data_message_hash;

fn digest(&self) -> Sha256Sum {
    let vote_message = TupleData::from_data(vec![("sip".into(), ClarityValue::UInt(self.sip.into())), ("vote", ClarityValue::UInt(self.vote.to_u8().into()))]).unwrap();
    let data_domain = make_structured_data_domain("signer-sip-voting", "1", CHAIN_ID_MAINNET);
    structured_data_message_hash(vote_message.into(), data_domain)
}
```

Reply:
Ah yes, this is how its done in generate stacking signature too. Also allows me to remove sha2 as a dependency. :D